### PR TITLE
Number of bin boundaries

### LINKED ISSUE #20495: MDNorm... algorithms dropping some points
When I run a script like 

```
w=Load(Filename='/SNS/HYS/IPTS-18741/nexus/HYS_174983.nxs.h5')
SetUB(Workspace=w, a=10.8, b=10.8, c=10.8, u='0,0,1', v='1,0,0')
dgs,_=DgsReduction(SampleInputWorkspace=w, SampleInputMonitorWorkspace=w, SofPhiEIsDistribution=False)
md=ConvertToMD(InputWorkspace=dgs, QDimensions='Q3D', 
               Q3DFrames='HKL', QConversionScales='HKL')
a,b=MDNormDirectSC(InputWorkspace=md, 
                   AlignedDim0='[0,0,L],0,5,200', 
                   AlignedDim1='[0,K,0],-.8,.8,100', 
                   AlignedDim2='[H,0,0],-4,1,1', 
                   AlignedDim3='DeltaE,-1,1,1')

cov=CalculateCoverageDGS(InputWorkspace=dgs, IncidentEnergy=18, 
                         Dimension1='Q3', Dimension1Min=0, Dimension1Max=5, Dimension1Step=0.025, 
                         Dimension2='Q2', Dimension2Min=-.8, Dimension2Max=.8, Dimension2Step=0.016, 
                         Dimension3='Q1', Dimension3Min=-4, Dimension3Max=1, Dimension3Step=5,
                         Dimension4='DeltaE', Dimension4Min=-1, Dimension4Max=1, Dimension4Step=2)
```

I expect that where I see non-zero pixels in the coverage image to get non-zero contributions in the normalization (b workspace). As you can see in the images below, I have some missing contribution around (0,1)

b
cov

### LINKED ISSUE #20497: Refactor `getNBins() + 1`
Create a function in `IMDDimension` to return the number of boundaries. Needs to be implemented in `MatrixWorkspace` (class `MWXDimension`) and in `MDHistoDimension`

### PR DESCRIPTION
**Pull request by AndreiSavici**
New function to get number of bin boundaries

Fixes #20497
Fixes #20495 (see PR #20496)

**Release Notes** 

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/src/SaveMD2.cpp [hunk 1]
```diff
@@ -133,10 +133,10 @@ void SaveMD2::doSaveHisto(Mantid::DataObjects::MDHistoWorkspace_sptr ws) {
   for (size_t d = 0; d < numDims; d++) {
     std::vector<double> axis;
     IMDDimension_const_sptr dim = ws->getDimension(d);
-    for (size_t n = 0; n < dim->getNBins() + 1; n++)
+    for (size_t n = 0; n < dim->getNBoundaries(); n++)
```
Comment (quantumsteve) [inline Framework/MDAlgorithms/src/SaveMD2.cpp]: Move `dim->getNBoundaries()` out of the for loop. Compilers have a hard time seeing through virtual functions.

File: qt/paraview_ext/VatesAPI/src/vtkMDHistoLineFactory.cpp [hunk 2]
```diff
@@ -73,10 +73,9 @@ vtkMDHistoLineFactory::create(ProgressAction &progressUpdating) const {
     const int nBinsX =
         static_cast<int>(m_workspace->getXDimension()->getNBins());
 
-    const coord_t maxX = m_workspace->getXDimension()->getMaximum();
     const coord_t minX = m_workspace->getXDimension()->getMinimum();
 
-    coord_t incrementX = (maxX - minX) / coord_t(nBinsX - 1);
+    coord_t incrementX = m_workspace->getXDimension()->getBinWidth();
```
Comment (quantumsteve) [inline qt/paraview_ext/VatesAPI/src/vtkMDHistoLineFactory.cpp]: :+1:

**Review by quantumsteve — CHANGES_REQUESTED** [review]
This looks good, just be careful with virtual functions inside tight loops.

**Review by quantumsteve — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
How does this interact with @peterfpeterson's PR introducing variable length histograms?

**Comment by martyngigg** [comment]
I think this should also go into `3.12` and not `3.11`

**Comment by AndreiSavici** [comment]
This should definitely be in 3.11. It fixes a bug that is causing problems for both diffuse elastic and direct geometry inelastic single crystal scattering. The extra stuff is just because @quantumsteve wanted more than a temporary fix. See issue #20495  and PR #20496

**Comment by quantumsteve** [comment]
@SimonHeybrock This adds a member function to `IMDDimension` to avoid off-by-one errors in existing code. If there is any overlap with variable length histograms, it should have already been addressed when earlier when creating this object.

**Pull request merged**